protocols/v2/roles-logic-sv2/src/channel_logic/proxy_group_channel.rs

```diff
@@ -0,0 +1,173 @@
+                merkle_path: m.merkle_path.clone().into_static(),
+                coinbase_tx_prefix: m.coinbase_tx_prefix.clone().into_static(),
+                coinbase_tx_suffix: m.coinbase_tx_suffix.clone().into_static(),
+            };
+            group.on_new_extended_mining_job(cloned);
+        }
+    }
+    pub fn last_received_job_to_standard_job(
+        &mut self,
+        channel_id: u32,
+        group_id: u32,
+    ) -> Result<NewMiningJob<'static>, Error> {
+        match self.channels.get_mut(&group_id) {
+            Some(group) => group.last_received_job_to_standard_job(channel_id),
+            None => Err(Error::GroupIdNotFound),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct GroupChannel {
+    hom_downstreams: HashMap<u32, StandardChannel>,
+    future_jobs: Vec<NewExtendedMiningJob<'static>>,
+    last_prev_hash: Option<SetNewPrevHash<'static>>,
+    last_valid_job: Option<NewExtendedMiningJob<'static>>,
+    last_received_job: Option<NewExtendedMiningJob<'static>>,
+}
+
+impl GroupChannel {
+    fn new() -> Self {
+        Self {
+            hom_downstreams: HashMap::new(),
+            future_jobs: vec![],
+            last_prev_hash: None,
+            last_valid_job: None,
+            last_received_job: None,
+        }
+    }
+    fn on_channel_success_for_hom_downtream(
+        &mut self,
+        m: OpenStandardMiningChannelSuccess,
+    ) -> Result<Vec<Mining<'static>>, Error> {
+        let channel = StandardChannel {
+            channel_id: m.channel_id,
+            group_id: m.group_channel_id,
+            target: m.target.clone().into(),
+            extranonce: m.extranonce_prefix.clone().into(),
+        };
+        let channel_id = m.channel_id;
+        let mut res = vec![];
+        //res.push(Mining::OpenStandardMiningChannelSuccess(m));
+        for extended_job in &self.future_jobs {
+            let standard_job = extended_to_standard_job(
+                extended_job,
+                &channel.extranonce.clone().to_vec(),
+                channel.channel_id,
+                None,
+            )
+            .ok_or(Error::ImpossibleToCalculateMerkleRoot)?;
+            res.push(Mining::NewMiningJob(standard_job));
+        }
+
+        if let Some(new_prev_hash) = &self.last_prev_hash {
+            res.push(Mining::SetNewPrevHash(new_prev_hash.clone()))
+        };
+
+        if let Some(last_valid_job) = &self.last_valid_job {
+            let standard_job = extended_to_standard_job(
+                last_valid_job,
+                &channel.extranonce.clone().to_vec(),
+                channel.channel_id,
+                None,
+            )
+            .ok_or(Error::ImpossibleToCalculateMerkleRoot)?;
+            res.push(Mining::NewMiningJob(standard_job));
+        };
+
+        self.hom_downstreams.insert(channel_id, channel);
+
+        Ok(res)
+    }
+    fn update_new_prev_hash(&mut self, m: &SetNewPrevHash) {
+        while let Some(job) = self.future_jobs.pop() {
```

Comment:
This is iterating and popping the last off each time - which means that this will set `last_valid_job` to the first valid job - i think you want to use `pop_front` here right?

Reply:
is fine like that the last future is likely the one that we want to use